src/program.cpp

```diff
@@ -105,829 +63,148 @@ program::program(const program& p) { assign(p); }
 {
-    return std::find_if(
-               impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-                   return std::addressof(*ins) == std::addressof(x);
-               }) != impl->instructions.end();
+    return main_module.has_instruction(ins);
 }
 
-std::size_t program::size() const { return impl->instructions.size(); }
-instruction_ref program::begin() const { return impl->instructions.begin(); }
-instruction_ref program::end() const { return impl->instructions.end(); }
+std::size_t program::size() const { return main_module.size(); }
+instruction_ref program::begin() const { return main_module.begin(); }
+instruction_ref program::end() const { return main_module.end(); }
 
-std::vector<shape> program::get_output_shapes() const
-{
-    auto last_ins = impl->instructions.back();
-    if(last_ins.name() == "@return")
-    {
-        const auto& output_ins = last_ins.inputs();
-        std::vector<shape> output_shapes;
-        std::transform(output_ins.begin(),
-                       output_ins.end(),
-                       std::back_inserter(output_shapes),
-                       [](auto& ins) { return ins->get_shape(); });
-
-        return output_shapes;
-    }
-    // The else branch is to provide backward compatibility
-    else
-    {
-        return {last_ins.get_shape()};
-    }
-}
+std::vector<shape> program::get_output_shapes() const { return main_module.get_output_shapes(); }
 
-context& program::get_context() const { return impl->ctx; }
+context& program::get_context() const { return main_module.get_context(); }
 
-instruction_ref program::validate() const
-{
-    return std::find_if(impl->instructions.begin(),
-                        impl->instructions.end(),
-                        [&](const instruction& i) { return !i.valid(impl->instructions.begin()); });
-}
+instruction_ref program::validate() const { return main_module.validate(); }
 
-bool program::is_compiled() const { return not this->impl->target_name.empty(); }
+bool program::is_compiled() const { return main_module.is_compiled(); }
 
-void program::compile(const target& t, compile_options options)
-{
-    assert(this->validate() == impl->instructions.end());
-    assert(not this->is_compiled());
-    this->impl->target_name = t.name();
-    this->impl->ctx         = t.get_context();
-    if(enabled(MIGRAPHX_TRACE_COMPILE{}))
-        options.trace = tracer{std::cout};
-    options.trace(*this);
-    options.trace();
-    run_passes(*this, t.get_passes(this->impl->ctx, options), options.trace);
-    auto invalid = this->validate();
-    if(invalid != impl->instructions.end())
-    {
-        auto index = std::distance(impl->instructions.begin(), invalid);
-        MIGRAPHX_THROW("Invalid program from compilation at instruction " + std::to_string(index));
-    }
-    this->finalize();
-}
+void program::compile(const target& t, compile_options options) { main_module.compile(t, options); }
```

Comment:
The module class probably shouldn't have a compile method, but we can leave this for now.